_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,416 @@
+
+![Share group](/assets/images/posts/2025-06-24-kafka-queue-04.png)
+
+As you can see, all the consumers are assigned to the same partitions and are fetching the records in parallel. Let’s describe the share group using the <code>kafka-share-groups.sh</code> console tool:
+
+![Share group describe](/assets/images/posts/2025-06-24-kafka-queue-05.png)
+
+There are 3 members in this share group and when describing the group without the <code>–-members</code> flag, the share partitions with their start offset is are listed. 
+
+Now let’s look at the regular consumer group (note that I am using the new consumer group protocol, not classic). 
+
+![Consumer group](/assets/images/posts/2025-06-24-kafka-queue-06.png)
+
+Only kafka-consumer-2 and kafka-consumer-1 are actively fetching from one partition each but kafka-consumer-0 is idle, not fetching records. Let’s describe this consumer group:
+
+![Consumer group describe](/assets/images/posts/2025-06-24-kafka-queue-07.png)
+
+There are 3 members in this group, but only 2 of them are assigned to a partition each exclusively as the topic has only 2 partitions. 
+
+### KafkaShareConsumer API
+
+Let’s look at the new Java API added for share group consumers. It looks very similar to KafkaConsumer API which makes it easier to use the new API if you are already familiar with it. With the Java API, users can do more fine grained acknowledgements of the records that are consumed and processed. There are 2 different mechanisms to acknowledge records:
+
+<i>Acknowledging records in batches</i>
+
+When acknowledging a batch of records, the implicit acknowledgment is used by calling the <code>poll()</code>. When fetching another batch of records by calling <code>poll()</code>, the batch of records delivered in the previous poll are marked as successfully processed and acknowledged. This is the simplest and the most efficient way to acknowledge records. In the following example, I’ve created a simple share consumer client that implicitly acknowledges batches of records:
+
+![KafkaShareConsumer API](/assets/images/posts/2025-06-24-kafka-queue-08.png)
+
+<i>Acknowledging individual records</i>
+
+This allows users to acknowledge an individual record depending on the outcome of the processing of the record. Each record is acknowledged using a call to <code>acknowledge()</code> which takes different types of acknowledgement as an argument: ACCEPT, RELEASE and REJECT.
+This aligns with the actions that can be taken by share group consumers mentioned previously.
+
+In the next example, I have used the new consumer configuration added for this feature, <b>share.acknowledgement.mode</b> and set it to "explicit". This configuration is set to "implicit" by default, which is why I didn’t need to set this configuration for the previous example. 
```

Comment:
>  which is why I didn’t need to set this configuration for the previous example. 

But I still see you set `share.acknowledgement.mode=explicit` in the previous example?

Reply:
That is a mistake, thanks for catching that. I have removed it now.